### PR TITLE
Add ControllerFailedToStart signature

### PR DESCRIPTION
**Pull request by eranco74**
It seems that sometimes the installation fails because the installer pod fail to be ready.
This is the first version of this signature.
See example signature [here](https://issues.redhat.com/browse/AITRIAGE-4917)
It should be extended with the reason that prevented the pod from starting/reaching a ready status.

### CODE REVIEW

File: tools/add_triage_signature.py [hunk 1]
```diff
@@ -2399,6 +2399,61 @@ def _process_ticket(self, url, issue_key):
...
+        try:
+            # Fetch pods.json from the bootstrap log bundle
+            pods_json = triage_logs_tar.get(self.PODS_PATH)
+        except FileNotFoundError:
+            return
+
+        # Fetch assisted-controller status
+        try:
+            pods = json.loads(pods_json)
+            controller_pod = [pod for pod in pods["items"] if pod["metadata"]["namespace"] == "assisted-installer"][0]
```
Comment (omertuc) [inline tools/add_triage_signature.py:2435]: Let's make another signature that does a generic dump table of all unhappy pods in that JSON, and their containerStatuses? Might be more revealing of a widespread issue than looking at just our container

File: tools/add_triage_signature.py [hunk 2]
```diff
@@ -2399,6 +2399,61 @@ def _process_ticket(self, url, issue_key):
...
+
+    def _process_ticket(self, url, issue_key):
+        md = get_metadata_json(url)
+        cluster_hosts = md["cluster"]["hosts"]
+        bootstrap_node = [host for host in cluster_hosts if host["bootstrap"]][0]
+
+        if bootstrap_node["progress"]["current_stage"] != "Waiting for controller":
+            return
+
+        triage_logs_tar = get_triage_logs_tar(triage_url=url, cluster_id=get_metadata_json(url)["cluster"]["id"])
```
Comment (omertuc) [inline tools/add_triage_signature.py:2425]: Don't download metadata twice

```suggestion
        cluster = get_metadata_json(url)["cluster"]
        bootstrap_node = [host for host in cluster["hosts"] if host["bootstrap"]][0]

        if bootstrap_node["progress"]["current_stage"] != "Waiting for controller":
            return

        triage_logs_tar = get_triage_logs_tar(triage_url=url, cluster_id=cluster["id"])
```

File: tools/add_triage_signature.py (hunk 1 above)
Comment (omertuc) [inline tools/add_triage_signature.py:2435]: Let's have more graceful handling of the controller pod `[0]` not existing at all

File: tools/add_triage_signature.py [hunk 3]
```diff
@@ -2399,6 +2399,61 @@ def _process_ticket(self, url, issue_key):
...
+            # Fetch pods.json from the bootstrap log bundle
+            pods_json = triage_logs_tar.get(self.PODS_PATH)
+        except FileNotFoundError:
+            return
+
+        # Fetch assisted-controller status
+        try:
+            pods = json.loads(pods_json)
+            controller_pod = [pod for pod in pods["items"] if pod["metadata"]["namespace"] == "assisted-installer"][0]
+            ready = [
```
Comment (omertuc) [inline tools/add_triage_signature.py:2436]: Also more graceful handling of "status", "conditions", "type", and the Ready condition not being present

File: tools/add_triage_signature.py [hunk 4]
```diff
@@ -2399,6 +2399,61 @@ def _process_ticket(self, url, issue_key):
...
+                The controller pod isn't ready, here is it's status:
+                h3. conditions:
+                {self._generate_table_for_report(controller_pod["status"]["conditions"])}
+                h3. containerStatuses:
+                {self._generate_table_for_report(controller_pod["status"]["containerStatuses"])}
+                """
+                    )
+                )
+        except Exception:
+            self._update_triaging_ticket("Failed to get controller pod from pods.json")
```
Comment (omertuc) [inline tools/add_triage_signature.py:2454]:
```suggestion
        except Exception as e:
            self._update_triaging_ticket("Failed to get readiness of controller pod from installer-gather resources/pods.json: {e}")
```

Include the error in the sig

File: tools/add_triage_signature.py (hunk 1 above)
Comment (eranco74) [reply tools/add_triage_signature.py:2435]: In another PR...

File: tools/add_triage_signature.py [hunk 5]
```diff
@@ -2399,6 +2399,72 @@ def _process_ticket(self, url, issue_key):
...
+            except IndexError:
+                raise "Failed to find controller pod in pods json"
+            try:
+                ready = [
+                    condition.get("status") == "True"
+                    for condition in controller_pod.get("status", {}).get("conditions", {})
+                    if condition.get("type") == "Ready"
+                ][0]
+            except IndexError:
+                raise f"Failed to determine controller pod in ready condition, here is it's status: {controller_pod['status']}"
```
Comment (osherdp) [inline tools/add_triage_signature.py:2449]:
```suggestion
                raise RuntimeError(f"Failed to determine controller pod in ready condition, here is it's status: {controller_pod['status']}")
```

File: tools/add_triage_signature.py [hunk 6]
```diff
@@ -2399,6 +2399,72 @@ def _process_ticket(self, url, issue_key):
...
+            if not ready:
+                self._update_triaging_ticket(
+                    dedent(
+                        f"""
+                The controller pod isn't ready, here is it's status:
```
Comment (osherdp) [inline tools/add_triage_signature.py:2454]:
```suggestion
                The controller pod isn't ready, here is its status:
```

**Review by osherdp — CHANGES_REQUESTED** [review]

File: tools/add_triage_signature.py [hunk 7]
```diff
@@ -2399,6 +2399,72 @@ def _process_ticket(self, url, issue_key):
...
+                raise "Failed to find controller pod in pods json"
+            try:
+                ready = [
+                    condition.get("status") == "True"
+                    for condition in controller_pod.get("status", {}).get("conditions", {})
+                    if condition.get("type") == "Ready"
+                ][0]
+            except IndexError:
+                raise f"Failed to determine controller pod in ready condition, here is it's status: {controller_pod['status']}"
+            if not ready:
```
Comment (osherdp) [inline tools/add_triage_signature.py:2450]: would it be beneficial to also output in case of success? (controller is in ready status)

**Review by osherdp — CHANGES_REQUESTED** [review]

File: tools/add_triage_signature.py (hunk 7 above)
Comment (eranco74) [reply tools/add_triage_signature.py:2450]: :+1:

Comment (eranco74) [reply tools/add_triage_signature.py]: Done

**Review by osherdp — APPROVED** [review]

**Pull request merged**